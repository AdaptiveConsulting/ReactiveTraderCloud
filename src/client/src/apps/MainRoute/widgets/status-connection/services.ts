import { compositeStatusService } from 'apps/MainRoute/store/singleServices'
import { map } from 'rxjs/operators'
export const allServiceStatus$ = compositeStatusService.serviceStatusStream.pipe(
  map(status => Object.values(status))
)
