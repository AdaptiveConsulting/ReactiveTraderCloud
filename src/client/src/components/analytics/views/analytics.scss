@import '../../common/styles/_theme';
@import '../../common/styles/_layout';

$positive-color: green;
$negative-color: $brand-danger;
$stroke-color: slategray;

.analytics-container {
  margin-left: 4px;
  align-self: stretch;
  overflow-x: hidden;
  overflow-y: auto;
  right: 0px;
  max-width: 400px;
  min-width: 350px;
  background: $gray-dark;
  box-shadow: -1px 0 1px rgba(0, 0, 0, .5), inset 0 0 1px rgba(255, 255, 255, .2);

  .header {
    font-size: 26px;
    font-weight: bold;
    color: white;
    padding-left: 8px;
    margin-right: 28px;
    .tearoff &.header {
      -webkit-app-region: drag;
    }
    small {
      color: $gray-light;
      font-size: .5em;
    }
  }

  .nv-container {
    position: relative;
    padding: 10px;
    border-top: 1px solid black;
    border-bottom: 1px solid black;
    box-shadow: -1px 0 1px rgba(0, 0, 0, .5), inset 0 0 1px rgba(255, 255, 255, .2);
    background: $gray-dark;
    &.negative {
      .nvd3 g.nv-groups g.nv-group .nv-area {

      }
    }
    .nvd3 g.nv-groups  {
      path.nv-line {
      }
      path.nv-area {

      }
      .nv-bar {
        rect {
          fill-opacity: .3;
          stroke: $stroke-color;
          stroke-width: .8;
          stroke-opacity: .4;
        }
        &.positive {
          fill: $positive-color;
          text {
            fill: $gray-light;
          }
        }
        text {
          font-size: 8px;
          fill: $gray-lighter;
          font-weight: normal;
        }
        &.negative {
          fill: $negative-color;
        }
        &:hover rect {
          stroke-opacity: .5;
          fill-opacity: .7;
        }
      }
    }

    .nv-controlsWrap.nvd3-svg, .nv-axisMin-y {
      display: none;
    }

    .nv-lineChart {
      .nv-axis.nv-y {
        text {
          font-size: 8px;
        }
      }

      .nv-axis.nv-x {
        text {
          font-size: 8px;
        }
      }
    }
  }

  .buttons {
    float: right;
    height: 38px;
    overflow: hidden;
    $shadow-width: 14px;
    button.btn {
      border-left: 1px solid black;
      color: $gray-light;
      box-sizing: border-box;
      outline: none !important;
      &.selected {
        position: relative;
        z-index: 100;
        color: $gray-lighter;
        background: linear-gradient(rgba(0, 0, 0, 0), $gray-darker) !important;
        box-shadow: -1px 0 1px rgba(0, 0, 0, .5), inset 0 0 1px rgba(255, 255, 255, .2);
      }
      &.selected:before {
        content: "";
        display: block;
        position: absolute;
        width: $shadow-width;
        height: 6px;
        transition: transform 1s;
        border-radius: 50%;
        left: 50%;
        transform: translate(-($shadow-width / 2), 0);
        top: -8px;
        box-shadow: 0 0 6px 4px rgba(255, 255, 255, .5);
      }
    }

  }
  @import 'nvd3';
}

.new-chart-area{
  stroke-width: 2px !important;
  fill: url('#areaGradient');
}
.new-chart-stroke{
  stroke: url('#chartStrokeLinearGradient');
}
