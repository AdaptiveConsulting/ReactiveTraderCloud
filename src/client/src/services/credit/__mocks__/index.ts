import { Observable, defer, EMPTY, of } from "rxjs"
import { RfqDetails } from "../creditRfqs/types"

let _creditRfqsById$: Observable<Observable<Record<number, RfqDetails>>>

export const creditRfqsById$ = defer(() => _creditRfqsById$)

export const __creditRfqsById = (
  input: Observable<Observable<Record<number, RfqDetails>>>,
) => {
  _creditRfqsById$ = input
}

export const __resetMocks = () => {
  _creditRfqsById$ = EMPTY
}

export const isBlotterDataStale$ = of(false)
