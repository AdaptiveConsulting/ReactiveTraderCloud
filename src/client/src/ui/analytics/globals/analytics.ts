import { injectGlobal } from 'emotion'
import {
  barBgColor,
  barchartTextColor,
  barHeight,
  negativeColor,
  pointerColor,
  positiveColor,
  transparentColor
} from './variables'

export default injectGlobal`
  .analytics__container {
    align-self: stretch;
    overflow-x: hidden;
    overflow-y: auto;
    height: 100%;
    padding: 0 1rem;

    #popout-content-container & {
      height: 100%;
    }
  }

  .analytics__barchart-indicator--negative {
    background-color: ${negativeColor};
  }

  .analytics__barchart-indicator--positive {
    background-color: ${positiveColor};
  }

  .analytics__controls {
    float: right;
  }

  .analytics__header {
    font-size: 1rem;
    padding-bottom: 0.5rem;
  }

  .analytics__header-value {
    font-weight: bold;
  }

  .analytics__header-title {
    display: block;
    margin-bottom: 1rem;
    color: currentColor;
  }

  .analytics__header-value--negative {
    color: ${negativeColor};
  }

  .analytics__header-value--positive {
    color: ${positiveColor};
  }

  .analytics__header-title-icon {
    margin-right: 17px;
  }

  .analytics__chart-container {
    position: relative;
    background: inherit;
    width: 100%;

    font-size: 0.75rem;
    .nv-lineChart {
      .nv-axis.nv-y {
        text {
          font-size: 0.5rem;
          fill: ${barchartTextColor};
        }
      }

      .nv-axis.nv-x {
        text {
          font-size: 0.5rem;
          fill: ${barchartTextColor};
        }
      }
    }
  }

  .analytics__barchart-container {
    vertical-align: middle;
    padding-top: 2px;
    padding-bottom: 0.75rem;
    height: 50px;
    &:hover {
    }
  }

  .analytics__barchart-bar {
    height: ${barHeight};
    overflow: hidden;
  }

  .analytics__barchart-bar-background {
    height: ${barHeight};
    width: 100%;
    background-color: ${barBgColor};
    position: absolute;
    margin-top: 3px;
  }

  .analytics__barchart-bar-wrapper {
    vertical-align: middle;
    height: 0.5rem;
    position: absolute;
    width: 100%;
    margin-top: 30px;
  }

  .analytics__barchart-border {
    height: 0.5rem;
    width: 2px;
    background-color: currentColor;
  }

  .analytics__barchart-border--center {
    position: absolute;
    left: calc(50% - 2px);
  }

  .analytics__barchart-border--left {
    float: left;
    display: inline;
  }

  .analytics__barchart-border--right {
    right: 0;
    position: absolute;
  }

  .analytics__barchart-title-wrapper {
    width: 100%;
    height: 30px;
    position: absolute;

    &:hover {
      .analytics__barchart-pointer-container {
        transform: scale(1.5);
      }

      .analytics__barchart-label {
        font-size: 0.75rem;
        cursor: default;
      }

      .analytics__barchart-label-amount {
        display: none;
      }
      .analytics__barchart-label-amount--hover {
        display: inline;
      }

      .analytics__barchart-label-currency-terms {
        display: none;
      }
    }
  }

  .analytics__barchart-pointer-container {
    position: absolute;
    margin-top: 1.5rem;
    transition: left linear $transition-duration;
    z-index: 99;
  }

  .analytics__barchart-pointer {
    width: 0;
    height: 0;
    border-width: 0.5rem 0.25rem 0 0.25rem;
    border-color: ${pointerColor} transparent transparent transparent;
    border-style: inset;
    transform: rotate(360deg);
    position: absolute;
    z-index: 100;
    left: -6px;
    margin-top: 2px;
  }

  .analytics__barchart-pointer-icon {
    text-shadow: 0px 0px 0 $pointer-color, -1px -1px 0 $pointer-color, 1px -1px 0 $pointer-color,
      -1px 1px 0 $pointer-color, 1px 1px 0 $pointer-color;
    rotation: 90deg;
  }

  .analytics__barchart-pointer--outline {
    position: absolute;
    width: 0;
    height: 0;
    z-index: 99;
    left: -0.5rem;
    border-width: 0.75rem 0.5rem 0 0.5rem;
    border-color: ${pointerColor} transparent transparent transparent;
    border-style: inset;
    transform: rotate(360deg);
  }

  .analytics__barchart-label {
    font-size: 13px;
    margin-top: 3px;
    white-space: nowrap;
  }

  .analytics__barchart-label-amount {
  }

  .analytics__barchart-label-amount--hover {
    display: none;
  }

  .analytics__barchart-label-pusher {
    height: 0.5rem;
    display: inline-block;
    transition: all linear $transition-duration;
  }
  .analytics__barchart-label-wrapper {
    transition: left linear $transition-duration;
    position: absolute;
    width: 100%;
    white-space: nowrap;
  }

  .analytics__barchart-container .analytics__barchart-amount {
    visibility: hidden;
    font-size: 18px;
    display: inline;
    float: right;
  }

  .analytics__barchart-container:hover .analytics__barchart-amount {
    visibility: visible;
  }

  .analytics__bubblechart-container {
    text-anchor: middle;
    height: 18rem;
  }

  .analytics__chart-title {
    font-size: 1rem;
  }

  .analytics__bubblechart-title {
    position: absolute;
  }

  .analytics__bubblechart-container {
    display: block;
    position: relative;
    padding-bottom: 2px;
  }

  .analytics__positions-label {
    fill: currentColor;
    font-weight: bold;
    font-size: 12px;
    pointer-events: none;
    select: none;
  }

  .analytics__positions-tooltip {
    position: absolute;
    width: auto;
    height: auto;
    padding: 2px 0.5rem;
    font-size: 0.75rem;
    background-color: currentColor;
    opacity: 1;
    box-shadow: 0.25rem 0.25rem 0.5rem rgba(0, 0, 0, 0.4);
    pointer-events: none;
    z-index: 100;
  }

  .new-chart-area {
    stroke-width: 2px !important;
    fill: url('#areaGradient');
  }
  .new-chart-stroke {
    stroke: url('#chartStrokeLinearGradient');
  }

  .analytics__icon--tearoff {
    cursor: pointer;
  }

  .analytics__icon--tearoff--hidden {
    display: none;
    visibility: hidden;
  }

  .stop1,
  .lineStop1 {
    stop-color: ${positiveColor};
  }

  .stop1End,
  .lineStop1End {
    stop-color: ${positiveColor};
  }

  .stop2,
  .lineStop2 {
    stop-color: ${negativeColor};
  }

  .stop2End,
  .lineStop2End {
    stop-color: ${negativeColor};
  }

  /* axis */
  .analytics__container .nvd3 .nv-axis path,
  .analytics__container .nvd3 .nv-axis .tick.zero line {
    stroke: currentColor;
  }
  /* axis labels */
  .analytics__chart-container .nv-lineChart .nv-axis.nv-x text,
  .analytics__chart-container .nv-lineChart .nv-axis.nv-y text {
    fill: currentColor;
  }
  /* grid */
  .analytics__container .nvd3 .nv-axis line {
    stroke: ${transparentColor};
  }

  .analytics__chart-tooltip {
    font-weight: 500;
    display: block;
    mix-blend-mode: difference !important;
  }

  .analytics__chart-tooltip-date {
    font-weight: 600;
  }

  .analytics__container--disconnected {
    display: flex;
    height: 100%;
    align-items: center;
    justify-content: center;
    font-size: 1rem;
  }
`
