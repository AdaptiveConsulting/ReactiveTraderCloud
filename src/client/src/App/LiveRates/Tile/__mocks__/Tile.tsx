import { Subject } from "rxjs"
import { CurrencyPair } from "services/currencyPairs"

export const Tile = ({
  currencyPair,
  isAnalytics,
}: {
  currencyPair: CurrencyPair
  isAnalytics: boolean
}) => (
  <div data-testid={`tile-${currencyPair.symbol}`}>
    IsAnalytics: {isAnalytics ? "true" : "false"}
  </div>
)

export const tile$ = () => new Subject()
