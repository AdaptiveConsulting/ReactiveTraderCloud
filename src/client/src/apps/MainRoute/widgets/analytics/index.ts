import './globals'
export { AnalyticsContainer } from './AnalyticsContainer'
