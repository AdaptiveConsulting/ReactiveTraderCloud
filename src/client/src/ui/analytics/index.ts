import './globals'

export { default as AnalyticsContainer } from './AnalyticsContainer'
export { default as analyticsReducer } from './reducer'
export { default as analyticsServiceEpic } from './epics'
