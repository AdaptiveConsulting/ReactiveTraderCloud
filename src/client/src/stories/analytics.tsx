import * as React from 'react'
import { storiesOf } from '@storybook/react'

import { default as analyticsProps, pnlChartModel, positionsChartModel } from './analytics/analyticsProps'
import { Analytics, PositionsBubbleChart, AnalyticsBarChart } from '../ui/analytics/index'
import PNLChart from '../ui/analytics/PNLChart'
import { getCurrencyPairs } from './currencyPairs'
import { PricePoint } from '../ui/analytics/model/pnlChartModel'

storiesOf('Analytics', module)
  .add('Full panel', () =>
    <Analytics canPopout={false} {...analyticsProps} currencyPairs={getCurrencyPairs()} />)
  .add('Bubble Chart', () =>
    <div className="analytics__bubblechart-container">
      <span className="analytics__chart-title analytics__bubblechart-title">Positions</span>
      <PositionsBubbleChart data={positionsChartModel.seriesData} size={{ width: 400, height: 300 }}/>
    </div>)
  .add('P&L chart', () =>
    <div className="analytics analytics__container animated fadeIn">

      <PNLChart seriesData={pnlChartModel.seriesData as PricePoint[]}
                lastPos={pnlChartModel.lastPos}
                maxPnl={pnlChartModel.maxPnl}
                minPnl={pnlChartModel.minPnl}
                options={pnlChartModel.options}/>
    </div>)
  .add('Bar chart', () =>
    <div className="analytics__chart-container">
      <span className="analytics__chart-title">PnL</span>
      <AnalyticsBarChart chartData={positionsChartModel.seriesData} isPnL={true} currencyPairs={getCurrencyPairs()}/>
    </div>)
