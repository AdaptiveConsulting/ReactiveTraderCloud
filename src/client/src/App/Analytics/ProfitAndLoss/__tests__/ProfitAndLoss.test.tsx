//  import { Subscribe } from "@react-rxjs/core"
import { BehaviorSubject } from "rxjs"

import { HistoryEntry } from "@/services/analytics"
import { analyticsMock } from "@/services/analytics/__mocks__/_analytics"
// import { TestThemeProvider } from "@/utils/testUtils"

// import { LineChart, lineChart$ } from "../LineChart"

vi.mock("@/services/analytics/analytics")

const historyMock1: HistoryEntry[] = [
  {
    timestamp: 1613402446096,
    usPnl: -2722245.789415593,
  },
  {
    timestamp: 1613402456097,
    usPnl: -2713309.3362381877,
  },
  {
    timestamp: 1613402466096,
    usPnl: -2700750.747100412,
  },
]

const historyMock2: HistoryEntry[] = [
  {
    timestamp: 1613402446096,
    usPnl: -2722245.789415593,
  },
  {
    timestamp: 1613402456097,
    usPnl: -2713309.3362381877,
  },
  {
    timestamp: 1613402466096,
    usPnl: -2700750.747100412,
  },
  {
    timestamp: 1613402466098,
    usPnl: -2700750.74710042,
  },
]

describe("Profit and Loss", () => {
  beforeEach(() => {
    analyticsMock.__resetMocks()
  })

  it.skip("renders the chart correctly", () => {
    const historyMock$ = new BehaviorSubject<HistoryEntry[]>(historyMock1)
    analyticsMock.__setHistoryMock(historyMock$)

    // const tree = renderer
    //   .create(
    //     <TestThemeProvider>
    //       <Subscribe source$={lineChart$} fallback="No data">
    //         <LineChart />
    //       </Subscribe>
    //     </TestThemeProvider>,
    //   )
    //   .toJSON()

    // expect(tree).toMatchSnapshot()
  })

  it.skip("renders the chart with updates", () => {
    const historyMock$ = new BehaviorSubject<HistoryEntry[]>(historyMock2)
    analyticsMock.__setHistoryMock(historyMock$)

    // const tree = renderer
    //   .create(
    //     <TestThemeProvider>
    //       <Subscribe source$={lineChart$} fallback="No data">
    //         <LineChart />
    //       </Subscribe>
    //     </TestThemeProvider>,
    //   )
    //   .toJSON()

    // expect(tree).toMatchSnapshot()
  })
})
