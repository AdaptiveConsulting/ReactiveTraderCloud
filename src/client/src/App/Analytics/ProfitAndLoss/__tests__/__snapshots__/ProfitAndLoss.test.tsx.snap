// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Profit and Loss renders the chart correctly 1`] = `"No data"`;

exports[`Profit and Loss renders the chart with updates 1`] = `"No data"`;
